File: 백준/Gold/2110. 공유기 설치/공유기 설치.cc
#include <iostream>
#include <algorithm>

using namespace std;

int n, c;
int arr[200020];

bool solve(int s){
    // 거리가 n이상이게 c개 설치할 수 있는지
    int cnt = 1;
    int prev = arr[0];
    
    for(int i = 1; i < n; i++){
        if(arr[i] - prev >= s){
            cnt++;
            prev = arr[i];
        }
    }
    
    return cnt >= c;
}

int main()
{

    cin >> n >> c;
    for(int i = 0; i < n;i++) cin >> arr[i];
    
    sort(arr, arr+n);
    
    int l = 1;
    int r = 1e9+1;
    
    while(l + 1 < r){
        int mid = (l+r) / 2;
        if(solve(mid)) l = mid;
        else r = mid;
    }
    
    cout << l;
    return 0;
}

File: 백준/Gold/2110. 공유기 설치/공유기 설치.java
import java.io.BufferedReader;
import java.io.FileInputStream;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.StringTokenizer;

public class Main {

    static int N, C;
    static int[] arr;

    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        StringTokenizer st = new StringTokenizer(br.readLine());
        N = Integer.parseInt(st.nextToken());
        C = Integer.parseInt(st.nextToken());
        arr = new int[N];

        for (int i = 0; i < N; i++) {
            arr[i] = Integer.parseInt(br.readLine());
        }

        Arrays.sort(arr);

        int s = 0;
        int e = arr[N - 1];

        int ret = -1;

        while (s <= e) {
            int mid = (s + e) / 2;
            int prv = 0;
            int cnt = 1;
            for (int i = 1; i < N; i++) {
                if (arr[i] - arr[prv] >= mid) {
                    prv = i;
                    cnt++;
                }
            }
            if (cnt < C) {
                e = mid - 1;
            } else {
                ret = Math.max(ret, mid);
                s = mid + 1;
            }
        }

        System.out.println(ret);
    }
}